src/wp-admin/includes/image.php

```diff
@@ -449,24 +571,155 @@ function _wp_make_subsizes( $new_sizes, $file, $image_meta, $attachment_id ) {
 		}
 	}
 
+	// Filter supported sizes for non primary mime types.
+	if ( ! $is_primary_mime ) {
+		$supported_multi_mime_sizes = _wp_multi_mime_get_supported_sizes( $attachment_id );
+		$new_sizes                  = array_filter(
+			$new_sizes,
+			function( $size ) use ( $supported_multi_mime_sizes ) {
+				return in_array( $size, $supported_multi_mime_sizes );
+			},
+			ARRAY_FILTER_USE_KEY
+		);
+	}
+
 	if ( method_exists( $editor, 'make_subsize' ) ) {
 		foreach ( $new_sizes as $new_size_name => $new_size_data ) {
-			$new_size_meta = $editor->make_subsize( $new_size_data );
+			$new_size_meta = $editor->make_subsize( $new_size_data, true );
 
 			if ( is_wp_error( $new_size_meta ) ) {
 				// TODO: Log errors.
 			} else {
 				// Save the size meta value.
-				$image_meta['sizes'][ $new_size_name ] = $new_size_meta;
+				if ( ! isset( $image_meta['sizes'][ $new_size_name ] ) ) {
+					$image_meta['sizes'][ $new_size_name ] = $new_size_meta;
+				} else {
+					// Remove any newly generated images that are larger than the primary mime type.
+					$new_size     = isset( $new_size_meta['filesize'] ) ? $new_size_meta['filesize'] : 0;
+					$primary_size = isset( $image_meta['sizes'][ $new_size_name ]['filesize'] ) ? $image_meta['sizes'][ $new_size_name ]['filesize'] : 0;
+
+					if ( $new_size && $new_size >= $primary_size ) {
+						wp_delete_file( dirname( $file ) . '/' . $new_size_meta['file'] );
+						continue;
+					}
+				}
+				if ( ! isset( $image_meta['sizes'][ $new_size_name ]['sources'] ) ) {
+					$image_meta['sizes'][ $new_size_name ]['sources'] = array();
+				}
+				$image_meta['sizes'][ $new_size_name ]['sources'][ $mime_type ] = _wp_get_sources_from_meta( $new_size_meta );
 				wp_update_attachment_metadata( $attachment_id, $image_meta );
 			}
 		}
 	} else {
 		// Fall back to `$editor->multi_resize()`.
-		$created_sizes = $editor->multi_resize( $new_sizes );
+		$created_sizes = $editor->multi_resize( $new_sizes, true );
 
 		if ( ! empty( $created_sizes ) ) {
-			$image_meta['sizes'] = array_merge( $image_meta['sizes'], $created_sizes );
+			foreach ( $created_sizes as $created_size_name => $created_size_meta ) {
+
+				// Primary mime type is set in 'sizes' array.
+				if ( ! isset( $image_meta['sizes'][ $created_size_name ] ) ) {
+					$image_meta['sizes'][ $created_size_name ] = $created_size_meta;
+				} else {
+					// Remove any newly generated images that are larger than the primary mime type.
+					$new_size     = isset( $created_size_meta['filesize'] ) ? $created_size_meta['filesize'] : 0;
+					$primary_size = isset( $image_meta['sizes'][ $created_size_name ]['filesize'] ) ? $image_meta['sizes'][ $created_size_name ]['filesize'] : 0;
+
+					if ( $new_size && $new_size >= $primary_size ) {
```

Comment:
See my other comment:
```suggestion
					if ( $new_size && $primary_size && $new_size >= $primary_size ) {
```